### PR TITLE
Support tracking VOLATILE SQL functions as mutations. Closes #1514

### LINKED ISSUE #1514: extend support for custom SQL functions (mutations, other functions types)
1. Support for SQL functions in mutations
2. Support other function types

### PR DESCRIPTION
**Pull request by jberryman**
### Description

Support tracking `VOLATILE` functions as mutations (top-level fields under `mutation` root field). Support is added to the v2 API while remaining backwards compatible. See comments for API design decisions.

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Server
- [ ] Console
- [ ] CLI
- [x] Docs
- [ ] Community Content
- [ ] Build System
- [x] Tests
- [ ] Other (list it)

### Related Issues

https://github.com/hasura/graphql-engine/issues/1514

### Solution and Design

I re-used the machinery from support for tracking functions as queries and tried to do as little as possible.

### Steps to test and verify

The tests written here represent the extent that I've tested/tried this out, so you can run those.

### Limitations, known bugs & workarounds

The tests aren't particularly thorough; since so much of the queries code is re-used this seems all right to me.

Aggregations on the result set isn't implemented; it might be as simple as uncommenting the relevant line (I just noticed this as an unresolved question).

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog 

#### Metadata

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [ ] Not required

#### GraphQL
- [x] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   
   

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     

  2. GraphQL API

     Schema Generation:
     
     
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     

     Schema Resolve:-
     
     - [ ] Change in treatment of `null` value for any input fields 
     

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed

### CODE REVIEW

**Comment by srghma** [comment]
Heterogeneous execution was merged https://github.com/hasura/graphql-engine/pull/5869

File: server/src-lib/Hasura/RQL/DDL/Schema/Function.hs
Comment (abooij) [inline server/src-lib/Hasura/RQL/DDL/Schema/Function.hs:210]: I don't understand the role of `Maybe` in `_fcAsMutation`. If it's only to derive instances of type classes, then let's obtain those instances in a different way (e.g. spelling them out, if needed).

It would also make sense to use the `RequestedFunction` type I proposed above here, so that we replace a boolean with an informative enumeration of options.

File: server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml
Comment (abooij) [inline server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml:62]: There's also a converse here, namely that non-volatile functions don't end up in the mutation root.

File: server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml
Comment (abooij) [inline server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml:48]: How should this TODO be resolved?

File: CHANGELOG.md
Comment (abooij) [inline CHANGELOG.md:81]: Personally I think an expanded changelog entry would be justified here as we are adding new API.

File: docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst
Comment (abooij) [inline docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst:151]:
```suggestion
     (i.e. with ``as_mutation: false``), while ``VOLATILE`` functions may *only* be exposed as mutations.
```

File: docs/graphql/core/schema/custom-functions.rst
Comment (abooij) [inline docs/graphql/core/schema/custom-functions.rst:38]:
```suggestion
 (i.e. with ``as_mutation: false``), while ``VOLATILE`` functions may *only* be exposed as mutations.
```

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs
Comment (abooij) [inline server/src-lib/Hasura/GraphQL/Execute/Query.hs:299]: I feel a bit iffy about adding a function, `convertFunction`, which is only used in `Hasura.GraphQL.Execute.Mutation`, but defined in `Hasura.GraphQL.Execute.Query`. It suggests there should be a module called `Hasura.GraphQL.Execute.Common`. Or is there a different way to resolve this?

Speaking of `convertFunction`, is it possible to have remote joins as part of a mutation custom function?

File: server/src-lib/Hasura/RQL/DDL/Schema/Function.hs
Comment (jberryman) [reply server/src-lib/Hasura/RQL/DDL/Schema/Function.hs:210]: > I don't understand the role of Maybe in _fcAsMutation

I considered this and I think the benefits of having consistent To/FromJSON instances outweigh the weirdness. We also are unconcerned with this in other parts of the codebase. I also don't love the situation though. Would be nice if there was a bit more control in aeson, so the serialization format doesn't leak so much into the data models

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/Query.hs:299]: > I feel a bit iffy ...

Yeah, I think it would be better to leave that for a separate PR, post- https://github.com/hasura/graphql-engine/pull/5797/files#diff-563b07c689cd5e65fabcd9d3915cb42bb1cc065d507af4ded924dba3fca9408c 

> Speaking of convertFunction, is it possible to have remote joins as part of a mutation custom function?

Do you mean "does that work in the PR?" or "is it possible to make it work?" ? I don't know the answer to either, but can look into it. Maybe a future enhancement?

File: server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml
Comment (jberryman) [reply server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml:48]: Ah right. I need to take a second look at this. Does it seem right to you?

File: server/src-lib/Hasura/GraphQL/Schema.hs
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Schema.hs:415]: After looking at this `buildGQLContext` is the only exported function here so I don't think this is a problem. The way I'd refactor this is to move it into the caller and have it closed over the partitioned list of `FunctionInfo`, but `Schema.hs` is getting a full rewrite in pull/5797 so I tried to touch as little as possible here

File: server/src-lib/Hasura/RQL/DDL/Schema/Function.hs
Comment (jberryman) [reply server/src-lib/Hasura/RQL/DDL/Schema/Function.hs:82]: I understand the boolean-blindness criticism here, but after trying a few things I just don't think it's worth it in this context, as this is all local to `Function.hs` and the provenance of the Bool is obvious.

File: server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml
Comment (jberryman) [reply server/tests-py/queries/graphql_mutation/functions/function_as_mutations.yaml:48]: Just turned this into a FIXME referencing https://github.com/hasura/graphql-engine/issues/6106 for now

**Comment by jberryman** [comment]
@abooij Thanks for the review. Hopefully not coming across as stubborn in how I came down on some of the suggestions here, though I did carefully (re)consider and appreciate them all. Let me know if you feel strongly about what remains open!

File: server/src-lib/Hasura/GraphQL/Schema.hs
Comment (abooij) [reply server/src-lib/Hasura/GraphQL/Schema.hs:426]: Okay, let's aim to write a clear docstring comment here for now!

File: server/src-lib/Hasura/RQL/DDL/Schema/Function.hs
Comment (abooij) [reply server/src-lib/Hasura/RQL/DDL/Schema/Function.hs:82]: Does this perhaps better describe this bit?

```haskell
data IntendedFunctionPosition = IFPQuery | IFPMutation deriving (Show, Eq)
```

File: server/src-lib/Hasura/RQL/DDL/Schema/Function.hs
Comment (abooij) [reply server/src-lib/Hasura/RQL/DDL/Schema/Function.hs:210]: I think in this case it would clarify the Haskell side of things. :-)

**Comment by jberryman** [comment]
@abooij please give a look over the last two commits. I'll rebase before merging. Thanks!

**Comment by jberryman** [comment]
@tirumaraiselvan can I get a changelog check please

File: CHANGELOG.md
Comment (tirumaraiselvan) [inline CHANGELOG.md:60]: typo `track_table`?

**Review by tirumaraiselvan — APPROVED** [review]
Changelog LGTM barring small typo.

File: docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst:56]:
```suggestion
Track as SQL function called ``search_articles`` with a Hasura session argument:
```

File: docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/api-reference/schema-metadata-api/custom-functions.rst:52]:
```suggestion
It supports more configuration options than v1, and also supports tracking
```

File: docs/graphql/core/schema/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/schema/custom-functions.rst:37]: There's something wrong here:

image

Probably an indentation error...

File: docs/graphql/core/schema/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/schema/custom-functions.rst:562]: Something wrong with the link here:

image

File: docs/graphql/core/schema/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/schema/custom-functions.rst:565]: Can we be specific what we mean by "these"?

**Review by abooij — APPROVED** [review]
Thanks, @jberryman!

File: docs/graphql/core/schema/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/schema/custom-functions.rst:569]:
```suggestion
   It's easy to accidentally give an SQL function the wrong volatility (or for a
```

File: docs/graphql/core/schema/custom-functions.rst
Comment (marionschleifer) [inline docs/graphql/core/schema/custom-functions.rst:571]: Can we explain in the text above how we can specify ``as_mutation``?

**Review by marionschleifer — CHANGES_REQUESTED** [review]

**Review by tirumaraiselvan — CHANGES_REQUESTED** [review]
Metadata spec review

**Comment by jberryman** [comment]
After discussion with @tirumaraiselvan we're going to rework this with a more permissive API wrt volatility:

- new field (replacing `as_mutation`): `expose_as: query` or `expose_as: mutation`
    - if included we'll place the function where requested, regardless of its volatility
    - ...with the exception that _non-volatile functions requested exposed as mutations will raise an error_, since this probably indicates a mistake either in choice of function, or when defining the function
    - we'll raise a warning in some way whenever the volatility and `expose_as` are "mismatched"
- if that field is omitted, we'll infer `expose_as` from the volatility (`VOLATILE` implies mutation, otherwise query)

**Comment by jberryman** [comment]
Mmm, seems I lost the race with fd8d51a37 ...

Since `Schema.hs` will need a full rewrite and we're changing the API and docs i'm going to close this and open a new PR that can be reviewed from scratch, which I think will be more clear.

**Comment by jberryman** [comment]
Progress here is blocked on some deeper review of `Schema.hs` changes in metadata separation PRs, see discussion in slack

**Comment by jberryman** [comment]
@marionschleifer thanks! I tried to incorporate those changes into v2 of this work, though there are a lot of changes

**Pull request closed without merging**
